icelake/src/io/mod.rs

```diff
@@ -17,22 +17,212 @@ pub use scan::*;
 use arrow_array::RecordBatch;
 use arrow_schema::SchemaRef;
 
-use crate::types::DataFileBuilder;
+use crate::types::{DataFileBuilder, StructValue};
 use crate::Result;
+pub mod input_wrapper;
+pub use input_wrapper::*;
+
+type DefaultInput = RecordBatch;
 
 #[async_trait::async_trait]
-pub trait RecordBatchWriter: Send + 'static {
-    async fn write(&mut self, batch: RecordBatch) -> Result<()>;
-    async fn close(&mut self) -> Result<Vec<DataFileBuilder>>;
+pub trait FileWriterBuilder: Send + Sync + Clone + 'static {
+    type R: FileWriter;
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R>;
 }
 
 #[async_trait::async_trait]
-pub trait RecordBatchWriterBuilder: Send + Sync + Clone + 'static {
-    type R: RecordBatchWriter;
+pub trait FileWriter: Send + 'static {
+    type R: FileWriteResult;
+    async fn write(&mut self, batch: &RecordBatch) -> Result<()>;
+    async fn close(self) -> Result<Self::R>;
+}
+
+pub trait FileWriteResult: Send + 'static {
+    type R: IcebergWriteResult;
+    fn empty() -> Self;
+    fn combine(&mut self, other: Self);
+    fn to_iceberg_result(self) -> Self::R;
+}
+
+#[async_trait::async_trait]
+pub trait IcebergWriterBuilder: Send + Clone + 'static {
+    type R;
     async fn build(self, schema: &SchemaRef) -> Result<Self::R>;
 }
 
-pub trait SingletonWriter: RecordBatchWriter {
-    fn current_file(&self) -> String;
+#[async_trait::async_trait]
+pub trait IcebergWriter<I = DefaultInput>: Send + 'static {
+    type R: IcebergWriteResult;
+    async fn write(&mut self, input: I) -> Result<()>;
+    async fn flush(&mut self) -> Result<Self::R>;
+}
+
+pub trait IcebergWriteResult: Send + Sync + 'static {
+    fn empty() -> Self;
+    fn with_content(&mut self, content: crate::types::DataContentType) -> &mut Self;
+    fn with_equality_ids(&mut self, equality_ids: Vec<i32>) -> &mut Self;
+    fn with_partition(&mut self, partition_value: Option<StructValue>) -> &mut Self;
+    fn combine(&mut self, other: Self);
+    fn flush(&mut self) -> Self;
```

Comment:
Could you elaborate these method?

Reply:
The main problem is that I'm not quite clear about the use case of it. As I have suggested, it's better to split this pr into smaller prs so that we can get context of the api.